File: packages/guards/src/bias-gender.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface BiasGenderOptions {
  action: 'block' | 'warn';
}

const GENDERED_TITLES: [RegExp, string][] = [
  [/\bchairman\b/i, 'Use "chairperson"'],
  [/\bfireman\b/i, 'Use "firefighter"'],
  [/\bpoliceman\b/i, 'Use "police officer"'],
  [/\bstewardess\b/i, 'Use "flight attendant"'],
  [/\bmailman\b/i, 'Use "mail carrier"'],
  [/\bmanpower\b/i, 'Use "workforce"'],
  [/\bmanmade\b/i, 'Use "artificial/synthetic"'],
];

const STEREOTYPE_PATTERNS: RegExp[] = [
  /\bwomen\s+should\b/i,
  /\bmen\s+always\b/i,
  /\bmen\s+are\s+better\s+at\b/i,
  /\bwomen\s+are\s+too\s+emotional\b/i,
  /\bgirls\s+can'?t\b/i,
  /\bboys\s+don'?t\s+cry\b/i,
  /\bnot\s+a\s+job\s+for\s+women\b/i,
  /\blike\s+a\s+girl\b/i,
  /\bman\s+up\b/i,
];

export function biasGender(options: BiasGenderOptions): Guard {
  return {
    name: 'bias-gender',
    version: '0.1.0',
    description: 'Detects gender bias: gendered titles, stereotypes, assumption patterns',
    category: 'content',
    supportedStages: ['input', 'output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];

      for (const [re, suggestion] of GENDERED_TITLES) {
        if (re.test(text)) issues.push(`gendered-title: ${suggestion}`);
      }

      for (const pat of STEREOTYPE_PATTERNS) {
        if (pat.test(text)) {
          issues.push('stereotype');
          break;
        }
      }

      const triggered = issues.length > 0;

      return {
        guardName: 'bias-gender',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues, reason: 'Gender bias detected' } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/bias_gender.py
"""Gender bias detection guard."""
from __future__ import annotations

import re
import time

from open_guardrail.core import GuardResult

_GENDERED_TITLES: list[tuple[re.Pattern[str], str]] = [
    (re.compile(r"\bchairman\b", re.I), "chairperson"),
    (re.compile(r"\bfireman\b", re.I), "firefighter"),
    (re.compile(r"\bpoliceman\b", re.I), "police officer"),
    (re.compile(r"\bstewardess\b", re.I), "flight attendant"),
    (re.compile(r"\bmailman\b", re.I), "mail carrier"),
    (re.compile(r"\bmanpower\b", re.I), "workforce"),
    (re.compile(r"\bmanmade\b", re.I), "artificial"),
]

_STEREOTYPES: list[re.Pattern[str]] = [
    re.compile(r"\bwomen\s+should\b", re.I),
    re.compile(r"\bmen\s+always\b", re.I),
    re.compile(r"\bmen\s+are\s+better\s+at\b", re.I),
    re.compile(r"\bwomen\s+are\s+too\s+emotional\b", re.I),
    re.compile(r"\bgirls\s+can'?t\b", re.I),
    re.compile(r"\bboys\s+don'?t\s+cry\b", re.I),
    re.compile(r"\bnot\s+a\s+job\s+for\s+women\b", re.I),
    re.compile(r"\blike\s+a\s+girl\b", re.I),
    re.compile(r"\bman\s+up\b", re.I),
]


class _BiasGender:
    def __init__(self, *, action: str = "warn") -> None:
        self.name = "bias-gender"
        self.action = action

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []

        for pat, suggestion in _GENDERED_TITLES:
            if pat.search(text):
                issues.append(
                    f"gendered-title: Use '{suggestion}'"
                )

        for pat in _STEREOTYPES:
            if pat.search(text):
                issues.append("stereotype")
                break

        triggered = len(issues) > 0
        elapsed = (time.perf_counter() - start) * 1000
        return GuardResult(
            guard_name="bias-gender",
            passed=not triggered,
            action=self.action if triggered else "allow",
            message=(
                "Gender bias detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {"issues": issues} if triggered else None
            ),
        )


def bias_gender(
    *, action: str = "warn"
) -> _BiasGender:
    return _BiasGender(action=action)
